File: Core/deep_risk_analyzer.py
import logging
from typing import List
import json

logger = logging.getLogger(__name__)


class DeepRiskAnalyzer:
    """
    LLM-based Risk Analyzer for RAG retrieval.
    inspects retrieved documents for semantic contradictions
    before generation.
    """

    def __init__(self, llm_service=None):
        """
        :param llm_service: Service to make LLM calls.
                            If None, attempts to get from global state or dependency injection.
        """
        self.llm_service = llm_service

    async def analyze_conflicts(
        self, query: str, retrieved_chunks: List[str], model: str = "gpt-3.5-turbo"
    ) -> List[str]:
        """
        Analyze chunks for contradictions regarding the query.
        Returns a list of warning strings.
        """
        if not retrieved_chunks or len(retrieved_chunks) < 2:
            return []

        # Construction of the analysis prompt
        context_text = ""
        for i, chunk in enumerate(retrieved_chunks):
            context_text += f"[Document {i + 1}]: {chunk}\n\n"

        system_prompt = (
            "You are an expert Fact-Checker. "
            "Your task is to identify if there are any FACTUAL CONTRADICTIONS between the provided documents "
            "regarding the user's query.\n"
            "If Document A says 'X is 5' and Document B says 'X is 10', that is a contradiction.\n"
            "If one document provides more detail than another, that is NOT a contradiction.\n"
            "Return the result as a JSON object with a key 'conflicts' which is a list of strings describing the contradictions.\n"
            "If no contradictions found, return {'conflicts': []}."
        )

        user_prompt = f"User Query: {query}\n\nRetrieved Documents:\n{context_text}"

        try:
            # We assume self.llm_service has a method similar to chat_completion or generate
            # adapting to the existing RAG pipeline's service structure
            if self.llm_service:
                response = await self.llm_service.chat_completion(
                    messages=[
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_prompt},
                    ],
                    model=model,
                    temperature=0.0,
                    json_mode=True,
                )

                content = response.get("content", "{}")
                data = json.loads(content)
                return data.get("conflicts", [])
            else:
                logger.warning("DeepRiskAnalyzer: No LLM service provided, skipping analysis.")
                return []

        except Exception as e:
            logger.error(f"DeepRiskAnalyzer failed: {e}")
            return []

File: rust_output/src/deep_risk_analyzer.rs
use anyhow::{Result, Context};
use std::collections::HashMap;
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// LLM-based Risk Analyzer for RAG retrieval.
/// inspects retrieved documents for semantic contradictions
/// before generation.
#[derive(Debug, Clone)]
pub struct DeepRiskAnalyzer {
    pub llm_service: String,
}

impl DeepRiskAnalyzer {
    /// :param llm_service: Service to make LLM calls.
    /// If None, attempts to get from global state or dependency injection.
    pub fn new(llm_service: String) -> Self {
        Self {
            llm_service,
        }
    }
    /// Analyze chunks for contradictions regarding the query.
    /// Returns a list of warning strings.
    pub async fn analyze_conflicts(&mut self, query: String, retrieved_chunks: Vec<String>, model: String) -> Result<Vec<String>> {
        // Analyze chunks for contradictions regarding the query.
        // Returns a list of warning strings.
        if (!retrieved_chunks || retrieved_chunks.len() < 2) {
            vec![]
        }
        let mut context_text = "".to_string();
        for (i, chunk) in retrieved_chunks.iter().enumerate().iter() {
            context_text += format!("[Document {}]: {}\n\n", (i + 1), chunk);
        }
        let mut system_prompt = "You are an expert Fact-Checker. Your task is to identify if there are any FACTUAL CONTRADICTIONS between the provided documents regarding the user's query.\nIf Document A says 'X is 5' and Document B says 'X is 10', that is a contradiction.\nIf one document provides more detail than another, that is NOT a contradiction.\nReturn the result as a JSON object with a key 'conflicts' which is a list of strings describing the contradictions.\nIf no contradictions found, return {'conflicts': []}.".to_string();
        let mut user_prompt = format!("User Query: {}\n\nRetrieved Documents:\n{}", query, context_text);
        // try:
        {
            if self.llm_service {
                let mut response = self.llm_service::chat_completion(/* messages= */ vec![HashMap::from([("role".to_string(), "system".to_string()), ("content".to_string(), system_prompt)]), HashMap::from([("role".to_string(), "user".to_string()), ("content".to_string(), user_prompt)])], /* model= */ model, /* temperature= */ 0.0_f64, /* json_mode= */ true).await;
                let mut content = response.get(&"content".to_string()).cloned().unwrap_or("{}".to_string());
                let mut data = serde_json::from_str(&content).unwrap();
                data.get(&"conflicts".to_string()).cloned().unwrap_or(vec![])
            } else {
                logger.warning("DeepRiskAnalyzer: No LLM service provided, skipping analysis.".to_string());
                vec![]
            }
        }
        // except Exception as e:
    }
}
